File: src-kotlin/task-manager-domain/src/main/kotlin/com/example/taskmanager/domain/tasks/TaskRepository.kt
package com.example.taskmanager.domain.tasks

/**
 * Persistence port for [Task] aggregates.
 */
interface TaskRepository {
    fun save(task: Task): Task

    fun findById(id: TaskId): Task?

    fun findByStatus(status: TaskStatus): List<Task>

    fun findAll(): List<Task>

    fun findTodoTasks(): List<Task> = findByStatus(TaskStatus.TODO)

    fun findInProgressTasks(): List<Task> = findByStatus(TaskStatus.IN_PROGRESS)

    fun findCompletedTasks(): List<Task> = findByStatus(TaskStatus.DONE)

    fun findActiveTasks(): List<Task> = findAll().filter { it.status.isActive }

    fun deleteById(id: TaskId)

    fun existsById(id: TaskId): Boolean

    fun count(): Long

    fun countByStatus(status: TaskStatus): Long
}

File: src-swift/Sources/TaskManagerDomain/TaskRepository.swift
/// Persistence port for `Task` aggregates.
public protocol TaskRepository: AnyObject {
    @discardableResult
    func save(_ task: Task) -> Task

    func findByID(_ id: TaskID) -> Task?

    func findByStatus(_ status: TaskStatus) -> [Task]

    func findAll() -> [Task]

    func deleteByID(_ id: TaskID)

    func exists(_ id: TaskID) -> Bool

    func count() -> Int

    func count(status: TaskStatus) -> Int
}

public extension TaskRepository {
    func findTodoTasks() -> [Task] {
        findByStatus(.todo)
    }

    func findInProgressTasks() -> [Task] {
        findByStatus(.inProgress)
    }

    func findCompletedTasks() -> [Task] {
        findByStatus(.done)
    }

    func findActiveTasks() -> [Task] {
        findAll().filter { $0.status.isActive }
    }
}
